node/bft/src/helpers/pending.rs

```diff
@@ -128,12 +131,29 @@ impl<T: Copy + Clone + PartialEq + Eq + Hash, V: Clone> Pending<T, V> {
 
     /// Removes the callbacks for the specified `item` that have expired.
     pub fn clear_expired_callbacks_for_item(&self, item: impl Into<T>) {
+        let item = item.into();
         // Clear the callbacks that have expired.
-        if let Some(callbacks) = self.callbacks.lock().get_mut(&item.into()) {
+        if let Some(callbacks) = self.callbacks.lock().get_mut(&item) {
             // Fetch the current timestamp.
             let now = OffsetDateTime::now_utc().unix_timestamp();
             // Remove the callbacks that have expired.
-            callbacks.retain(|(_, timestamp)| now - *timestamp <= CALLBACK_TIMEOUT_IN_SECS);
+            callbacks.retain(|(_, timestamp)| now - *timestamp <= CALLBACK_EXPIRATION_IN_SECS);
+        }
+
+        // If there are no more remaining callbacks for the item, remove the item from the pending queue.
+        // If the item was added to the pending queue without a callback, do not remove the item.
+        let mut callbacks = self.callbacks.lock();
```

Comment:
we can obtain this guard just once, before clearing the expired callbacks, and it will be applicable for both blocks

Reply:
also, since we're querying the same item twice, we could do it with a single query with the `Entry` API to avoid borrowing issues (a matched `Entry::Occupied` can be removed from the collection)

Reply:
Good catch, I've made it a singular lock. I left the `get_mut` and `remove` approach rather than changing to the `Entry` API, but can easily be amended if `Entry` is a better approach.

Reply:
as long as the borrow checker is happy, `get_mut` is just fine :+1: